### ISSUE TITLE
Connection drops intermittently

### ISSUE THREAD

**Issue by aakkam22** [mikebrady/shairport-sync#992]
Hello I am having problem with shairport-sync and would appreciate your assistance:

I am running shairport-sync on a raspberry pi 3 model b. The rpi is connected to the lan via 100 mbps ethernet connection and both wifi and bt are disabled with a dtoverlay in /boot/config.txt. The router is a google fiber network box with 802.11ac.

I connect using an iPad Mini 5 and play youtube videos. Sometimes the connection drops. When I try to disconnect and reconnect using the airplay menu, I get different errors. Sometimes it was "speakers are being used by someone else" and other times it was "speakers are unavailable". 

I found a past issue here on the repo and changed the shairport-sync.conf file to allow session interrupts and change the session timeout from 120 seconds to 0, but while I don't see those specific errors anymore, the connection is still dropping.

I ran the shairport-sync with -vvv triple debug option. The log file exceeded the size limit for BASIC pastebin users so I uploaded it to google drive for your viewing pleasures. It seems that there are multiple connections being made and the iPad keeps sending teardown messages, but I do not know what the problem could be at this point and would appreciate some help. Please let me know if you would also like to see the config file. The only changes I have made are changing the hostname and output device to alsa eq (the equalizer I am using for the the output of the hifiberry) and as of new, the timeout and interrupt settings. Everything else is defaults. I am using the latest release of shairport-sync.

[log](https://drive.google.com/open?id=17fj20NHnsZfwS4JMtZFcPkXtDeYi5Yi0)

Thanks u so so much

**Comment by mikebrady** [comment]
Thanks for the post. Would you be kind enough to generate another log, this time with the verbosity at 2 but with statistics enabled? This gives an idea of packets missed and stuff like that.

Also, if you could capture Shairport Sync log entries from when it starts, that would be useful.

**Comment by aakkam22** [comment]
Sure. I have not noticed a pattern of when this happens or any correlations, it seems to be random, so I'll have to leave logging on until it happens.

**Comment by aakkam22** [comment]
I cleared the log and restarted the service, then connected. Playback lasted for almost 40 minutes before dropping out, you can see the timestamps from 9:54 to 10:32. 

[log](https://drive.google.com/open?id=1hH2CaJvh86fT87Z7ElDDmxhFPBgpp99s)

**Comment by mikebrady** [comment]
Many thanks. Would you be able to turn on the gathering of statistics please?

It's in the `diagnostics` part of the configuration file, or alternatively add it as `--statistics` on the command line.

**Comment by aakkam22** [comment]
Oops. I forgot to comment out the line in the config. Do you need another log?

**Comment by mikebrady** [comment]
If possible, please. And if you could leave is as long as you can bear :) that would be great.

**Comment by aakkam22** [comment]
Will do. Thank you :)

**Comment by aakkam22** [comment]
OK, here is another log hopefully this is the one you are looking for. Sorry for the delay been away for a bit and have not had a chance to use the system much. The disconnect happened at 10:51:55

[syslog](https://drive.google.com/open?id=1a2g9ZzOqFIJlIvn4Bchxh40DwffVipV5)

**Comment by mikebrady** [comment]
Thanks.

The log shows that some packets didn't make it to the Raspberry Pi, even though resends were requested. The question is where the loss occurred. Since the Pi is wired, one could guess that the loss could occur between the iPad and the router, or possibly even inside the router.

There is also a moderate level of variation in the longest interval between packets – up to 220 milliseconds. I've seen worse, TBH. For reference, the average should be just below 8 milliseconds. I have no knowledge of the Google Fiber box, I'm afraid. It might be worth checking for firmware updates for it. Another thing to check is whether there are any sources of interference like microwave ovens and suchlike in the vicinity.

Everything else looks peachy.

**Comment by aakkam22** [comment]
Do you think the placement of the router could be the issue? Like if it’s far away. Or even the placement of the iPad, for example if it’s far away, or near a window, or in direct sunlight, etc.?

**Comment by aakkam22** [comment]
I will need to do some investigating, might be the cabling used too. The cable might not be an official cable, maybe was crimped by a student in the lab I work in.

**Comment by mikebrady** [comment]
Actually, I slightly misread the logs. Let me explain:

The network is dropping some packets alright, but not a huge number, and – crucially – when Shairport Sync asked for them to be resent, they were indeed resent, and no packets were missing when they were needed for playing.

For example:
```
Apr 12 10:51:31 raspberrypi shairport-sync:          1.853853539 "player.c:2682"       2.00,     -39.7,      39.7,       20060,      0,     37,      0,      4,   8558,  209,  223,   44100.00,   44105.61,   44100.55,     56.35,    26,    -43.97
```
Notice the number after the 20060, (the 20060 is the total number of packets received). It is 0. This mean that no packets were missing when they were needed for playing. The next number, 37, is the number of resent packets that were received. The next number, 0, is the number of resent packets that arrived too late to be played, and the next number, 4, is the number of resend requests made (a single request can ask for a sequence of packets).

Overall, these statistics show that the network is a bit noisy, but is within limits, and Shairport Sync is able to cope with it. (My previous statement was incorrect about this – apologies. There is probably no benefit to be had playing around with the network itself – it seems adequate.)

The bigger picture is that the client (YouTube on the iPad?) is terminating the connection at 10:51:36:
```
Apr 12 10:51:36 raspberrypi shairport-sync:          5.272072198 "rtsp.c:2373" Connection 84: Received an RTSP Packet of type "TEARDOWN":
```
and then re-establishing a connection at 10:51:40:
```
Apr 12 10:51:40 raspberrypi shairport-sync:          0.081435652 "rtsp.c:2373" Connection 85: Received an RTSP Packet of type "ANNOUNCE":
```
and then terminating it at 10:51:55:
```
Apr 12 10:51:55 raspberrypi shairport-sync:          4.611692511 "rtsp.c:2373" Connection 85: Received an RTSP Packet of type "TEARDOWN":
```
So, it looks like the client is, for some reason, dropping the connection from time to time. Can you recall what was happening at the time. For example, was YouTube moving from one track to the next? If so, that might explain this behaviour.

**Comment by aakkam22** [comment]
I did some testing on the network and found out that the local speeds were quite poor. Averaging around 15 Mbps when they should be averaging 300-400 Mbps since both the router and the client are using 802.11ac. Yes, 15 Mbps on average should be sufficient for streaming and playback, but nonetheless it is underperforming. 

So there were two problems: first the router was set to use only the 2.4GHz channel, and second the wifi band was being automatically selected for band 11. I changed the band to band 6 which dramatically improved speeds, but then I also enabled the 5GHz channel. Now, on the 5GHz channel using band 36 I get speeds up to 500 Mbps, far better than 15 Mbps.

I don't know why the iPad is sending teardown requests. There doesn't seem to be a pattern. Sometimes it will be in the middle of a track and then quit, without touching anything. I will continue to monitor the situation after the changes I made and also watch the iPad for any patterns.

Thanks
